# Momento de Retroalimentación: Módulo 2 Implementación de una técnica de aprendizaje máquina sin el uso de un framework. (Portafolio Implementación) [semana 3]

**Mario Javier Soriano Aguilera A01384282**

In [62]:
from google.colab import drive
drive.mount("/content/gdrive")
!pwd

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).
/content


In [63]:
!cp /content/gdrive/MyDrive/InteligenciaArtificial/MachineLearning/LogReg/MachineLearning.py .

In [87]:
import pandas as pd
import numpy as np
import MachineLearning
from MachineLearning import LogisticRegression
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split

sns.set()

El data set a utilizar va ser el de

In [65]:
df = pd.read_csv('/content/gdrive/MyDrive/InteligenciaArtificial/MachineLearning/LogReg/Cancer_Data.csv')

In [66]:
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

Limpieza de datos. Se observa que mis datos tienen 33 columnas y 569 filas. Ademas de contener valores nulos en toda una columna que ni siquiera tiene nombre, la cual se eliminara por que no brinda valor alguno

In [67]:
len(df)

569

In [68]:
df.isna().sum()

id                           0
diagnosis                    0
radius_mean                  0
texture_mean                 0
perimeter_mean               0
area_mean                    0
smoothness_mean              0
compactness_mean             0
concavity_mean               0
concave points_mean          0
symmetry_mean                0
fractal_dimension_mean       0
radius_se                    0
texture_se                   0
perimeter_se                 0
area_se                      0
smoothness_se                0
compactness_se               0
concavity_se                 0
concave points_se            0
symmetry_se                  0
fractal_dimension_se         0
radius_worst                 0
texture_worst                0
perimeter_worst              0
area_worst                   0
smoothness_worst             0
compactness_worst            0
concavity_worst              0
concave points_worst         0
symmetry_worst               0
fractal_dimension_worst      0
Unnamed:

In [69]:
df.drop(['id', 'Unnamed: 32'], axis=1, inplace=True)

Se observa el tipo de dato y vemos que tenemos la variable diagnosis que es tipo object y todos los demas valores numericos float64, usaremos la variable categorica diagnosis como nuestro target de prediccion.

In [70]:
df.dtypes

diagnosis                   object
radius_mean                float64
texture_mean               float64
perimeter_mean             float64
area_mean                  float64
smoothness_mean            float64
compactness_mean           float64
concavity_mean             float64
concave points_mean        float64
symmetry_mean              float64
fractal_dimension_mean     float64
radius_se                  float64
texture_se                 float64
perimeter_se               float64
area_se                    float64
smoothness_se              float64
compactness_se             float64
concavity_se               float64
concave points_se          float64
symmetry_se                float64
fractal_dimension_se       float64
radius_worst               float64
texture_worst              float64
perimeter_worst            float64
area_worst                 float64
smoothness_worst           float64
compactness_worst          float64
concavity_worst            float64
concave points_worst

In [71]:
df.diagnosis.unique()

array(['M', 'B'], dtype=object)

In [72]:
#df['diagnosis'] = df['diagnosis'].map({'M':1,'B':0})
#df.head()

In [73]:
df.describe()

radius_mean  texture_mean  perimeter_mean    area_mean  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  fractal_dimension_mean  ...  radius_worst  \
count     569.000000              569.000000  ...    569.000000   
mean        0.181162                0.062798  ...     16.269190   
std         0.027414                0.007060  ...      4.833242   
min         0.106000                0.049960  ...      7.930000   
25%         0.161900                0.057700  ...     13.010000   
50%         0.179200                0.061540  ...     14.970000   
75%         0.195700                0.066120  ...     18.790000   
max         0.304000                0.097440  ...     36.040000   

       texture_worst  perimeter_worst   area_worst  smoothness_worst  \
count     569.000000       569.000000   569.000000        569.000000   
mean       25.677223       107.261213   880.583128          0.132369   
std         6.146258        33.602542   569.356993          0.022832   
min        12.020000        50.410000   185.200000          0.071170   
25%        21.080000        84.110000   515.300000          0.116600   
50%        25.410000        97.660000   686.500000          0.131300   
75%        29.720000       125.400000  1084.000000          0.146000   
max        49.540000       251.200000  4254.000000          0.222600   

       compactness_worst  concavity_worst  concave points_worst  \
count         569.000000       569.000000            569.000000   
mean            0.254265         0.272188              0.114606   
std             0.157336         0.208624              0.065732   
min             0.027290         0.000000              0.000000   
25%             0.147200         0.114500              0.064930   
50%             0.211900         0.226700              0.099930   
75%             0.339100         0.382900              0.161400   
max             1.058000         1.252000              0.291000   

       symmetry_worst  fractal_dimension_worst  
count      569.000000               569.000000  
mean         0.290076                 0.083946  
std          0.061867                 0.018061  
min          0.156500                 0.055040  
25%          0.250400                 0.071460  
50%          0.282200                 0.080040  
75%          0.317900                 0.092080  
max          0.663800                 0.207500  

[8 rows x 30 columns]

In [74]:
Tcorrelation = df.corr(method='pearson')
Tcorrelation

<ipython-input-74-6004b698a5e9>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  Tcorrelation = df.corr(method='pearson')


radius_mean  texture_mean  perimeter_mean  area_mean  \
radius_mean                 1.000000      0.323782        0.997855   0.987357   
texture_mean                0.323782      1.000000        0.329533   0.321086   
perimeter_mean              0.997855      0.329533        1.000000   0.986507   
area_mean                   0.987357      0.321086        0.986507   1.000000   
smoothness_mean             0.170581     -0.023389        0.207278   0.177028   
compactness_mean            0.506124      0.236702        0.556936   0.498502   
concavity_mean              0.676764      0.302418        0.716136   0.685983   
concave points_mean         0.822529      0.293464        0.850977   0.823269   
symmetry_mean               0.147741      0.071401        0.183027   0.151293   
fractal_dimension_mean     -0.311631     -0.076437       -0.261477  -0.283110   
radius_se                   0.679090      0.275869        0.691765   0.732562   
texture_se                 -0.097317      0.386358       -0.086761  -0.066280   
perimeter_se                0.674172      0.281673        0.693135   0.726628   
area_se                     0.735864      0.259845        0.744983   0.800086   
smoothness_se              -0.222600      0.006614       -0.202694  -0.166777   
compactness_se              0.206000      0.191975        0.250744   0.212583   
concavity_se                0.194204      0.143293        0.228082   0.207660   
concave points_se           0.376169      0.163851        0.407217   0.372320   
symmetry_se                -0.104321      0.009127       -0.081629  -0.072497   
fractal_dimension_se       -0.042641      0.054458       -0.005523  -0.019887   
radius_worst                0.969539      0.352573        0.969476   0.962746   
texture_worst               0.297008      0.912045        0.303038   0.287489   
perimeter_worst             0.965137      0.358040        0.970387   0.959120   
area_worst                  0.941082      0.343546        0.941550   0.959213   
smoothness_worst            0.119616      0.077503        0.150549   0.123523   
compactness_worst           0.413463      0.277830        0.455774   0.390410   
concavity_worst             0.526911      0.301025        0.563879   0.512606   
concave points_worst        0.744214      0.295316        0.771241   0.722017   
symmetry_worst              0.163953      0.105008        0.189115   0.143570   
fractal_dimension_worst     0.007066      0.119205        0.051019   0.003738   

                         smoothness_mean  compactness_mean  concavity_mean  \
radius_mean                     0.170581          0.506124        0.676764   
texture_mean                   -0.023389          0.236702        0.302418   
perimeter_mean                  0.207278          0.556936        0.716136   
area_mean                       0.177028          0.498502        0.685983   
smoothness_mean                 1.000000          0.659123        0.521984   
compactness_mean                0.659123          1.000000        0.883121   
concavity_mean                  0.521984          0.883121        1.000000   
concave points_mean             0.553695          0.831135        0.921391   
symmetry_mean                   0.557775          0.602641        0.500667   
fractal_dimension_mean          0.584792          0.565369        0.336783   
radius_se                       0.301467          0.497473        0.631925   
texture_se                      0.068406          0.046205        0.076218   
perimeter_se                    0.296092          0.548905        0.660391   
area_se                         0.246552          0.455653        0.617427   
smoothness_se                   0.332375          0.135299        0.098564   
compactness_se                  0.318943          0.738722        0.670279   
concavity_se                    0.248396          0.570517        0.691270   
concave points_se               0.380676          0.642262        0.683260   
symmetry_se                    

In [75]:
c = df.corr()
threshold = .80
np.abs(c.values) > threshold

<ipython-input-75-cd2d2f2a6e2b>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  c = df.corr()


array([[ True, False,  True,  True, False, False, False,  True, False,
        False, False, False, False, False, False, False, False, False,
        False, False,  True, False,  True,  True, False, False, False,
        False, False, False],
       [False,  True, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False,
        False, False, False,  True, False, False, False, False, False,
        False, False, False],
       [ True, False,  True,  True, False, False, False,  True, False,
        False, False, False, False, False, False, False, False, False,
        False, False,  True, False,  True,  True, False, False, False,
        False, False, False],
       [ True, False,  True,  True, False, False, False,  True, False,
        False, False, False, False,  True, False, False, False, False,
        False, False,  True, False,  True,  True, False, False, False,
        False, False, False],
       [False, False, False,

In [76]:
[f"{c.columns[i]} and {c.columns[j]}" for i, j in zip(*np.where(np.abs(c.values) > threshold)) if i < j]

['radius_mean and perimeter_mean',
 'radius_mean and area_mean',
 'radius_mean and concave points_mean',
 'radius_mean and radius_worst',
 'radius_mean and perimeter_worst',
 'radius_mean and area_worst',
 'texture_mean and texture_worst',
 'perimeter_mean and area_mean',
 'perimeter_mean and concave points_mean',
 'perimeter_mean and radius_worst',
 'perimeter_mean and perimeter_worst',
 'perimeter_mean and area_worst',
 'area_mean and concave points_mean',
 'area_mean and area_se',
 'area_mean and radius_worst',
 'area_mean and perimeter_worst',
 'area_mean and area_worst',
 'smoothness_mean and smoothness_worst',
 'compactness_mean and concavity_mean',
 'compactness_mean and concave points_mean',
 'compactness_mean and compactness_worst',
 'compactness_mean and concavity_worst',
 'compactness_mean and concave points_worst',
 'concavity_mean and concave points_mean',
 'concavity_mean and concavity_worst',
 'concavity_mean and concave points_worst',
 'concave points_mean and radius_wo

En lo anterior basicamente solo se realizo una matriz de correlacion ya que tenemos muchos labels y solo queremos valores significativos, de forma rapida se analizaron con una matris de correlacion, se realizo una matriz booleana de esa matriz de correlacion y se le pusieron las etiquetas para poder identificar a simple vista cuales son las variables y con cual otra tienen alta correlacion, en este caso se escogio correlacion de .7, y se escogieron las que mas se repetian ya que son las que mas variables pueden representar.

In [77]:
labels = df[['radius_mean', 'perimeter_mean', 'area_mean', 'compactness_mean', 'concave points_mean', 'texture_mean']]

In [81]:
X = labels
y = df['diagnosis'] = np.where(df['diagnosis']=='M',1,0)


In [88]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1234)

clf = LogisticRegression()
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)

def accuracy(yPredicted, yReal):
   return np.sum(yPredicted == yReal)/len(yReal)

acc = accuracy(y_pred, y_test)
print(acc)

0.8333333333333334


Se obtuvo un accuracy del .83 de forma obvia hay cuestiones a mejorar ya que se realizo un analisis superficial a los datos y seria buena idea realizar un analisis estadistico mas profundo sobre las variables a emplear en nuestro modelo para que pueda disernir mejor nuestros datos. Como por ejemplo ver como se distribuyen, boxplot, variables con colinealidad a nuestro modelo, pruebas de normalidad, tranformaciones, etc.

# Explicación del Modelo

El código (from MachineLearning import LogisticRegression )define una clase de regresión logística que tiene los siguientes métodos:

**sigmoid**: es una función que calcula el valor de la función sigmoide para un número x. La función sigmoide es una función que toma valores entre 0 y 1 y se usa para modelar la probabilidad de una clase binaria, en este caso M (maligno) y B (benigno) que son como se clasifican a los tipos de cancer dependiendo de sus caracteristicas y riesgos que pueda causar al que lo porta.

**init**: es el constructor de la clase que inicializa los atributos de la instancia, como la tasa de aprendizaje, el número de iteraciones, los pesos y el sesgo del modelo. Los pesos y el sesgo son los parámetros que se van a aprender mediante el ajuste del modelo a los datos, osea el fit.

**fit**: es el método que ajusta el modelo a los datos de entrenamiento X y y. Para ello, utiliza un algoritmo de descenso de gradiente que actualiza los pesos y el sesgo en cada iteración, minimizando la función de costo logística. El método también calcula la pérdida promedio en cada iteración. En este caso se estableció que los pesos empiecen en ceros.

**predict**: es el método que hace predicciones para nuevos datos X. Para ello, calcula la regresión lineal con los pesos y el sesgo aprendidos, y luego aplica la función sigmoide para obtener las probabilidades de cada clase. Luego, asigna una etiqueta de clase 0 o 1 según si la probabilidad es menor o mayor que 0.5, respectivamente. El método devuelve una lista de predicciones de clase.

**accuracy**: calcula la precisión de las predicciones de un modelo de clasificación, mide la cantidad de valores predecidos que fueron acertados con los reales, y se divide entre el numero de valores totales.

# ¿Por qué Regresion logistica?
En lo personal me interesaba aprender la formula y llevar acabo un modelo no tan simple pero tampoco tan complicado para poder llevarlo a practica, osea se escogio primero el modelo y luego se busco un dataset adecuado para poder hacer la implementacion.

# ¿Por qué se dice que el dataset es adecuado para Regresion Logistica?
Porque este tipo de modelo es utilizado cuando se quiere hacer una clasificacion binaria, osea mi variable respuesta tiene que ser categorica y binaria, en este caso cancer maligno o benigno, ademas de tener  labels numericos con los que el modelo de regresion logistica puede utilizar para relaizar mi clasificacion. Tambien puede hacer predicciones teniendo como variables explicativas categoricas, y en ese caso se tendrian que emplear tecnicas como variables dummies, las cuales indican la presencia o ausencia de cada categoria.
